PHP 8.1 / 8.2 Compatibility

swissspidy opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->

Since we use parts of the AMP plugin in the Web Stories plugin, it would be reassuring for us if it works more or less seamlessly with PHP 8.1 and 8.2.

PHP 8.2 compatibility is currently blocked by work on WordPress core itself.

Even PHP 8.1 is not seamless due to the conflict with the Requests library.

See also https://github.com/ampproject/amp-toolbox-php/pull/533 (for which an update to `dev-main` is included in this PR)

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

swissspidy committed:
Run tests against PHP 8.1

swissspidy committed:
Run tests against PHP 8.2
          - php: '8.2'
            wp: 'trunk'
            experimental: true

swissspidy committed:
Change php file to trigger CI

swissspidy committed:
Update PHPUnit depending on matrix

swissspidy committed:
Fix setup/teardown method names in `BlockUniqidTransformerTest`

swissspidy committed:
Add `ReturnTypeWillChange` attribute in `ElementList`

swissspidy committed:
Remove conflicting Requests library

swissspidy committed:
Remove float key test case in `EventTest`

Using floats as array keys is deprecated in PHP 8.1

swissspidy committed:
ignore platform reqs

swissspidy committed:
Add `file_exists` check

swissspidy committed:
Don’t convert to exceptions for now

swissspidy committed:
Install PHPUnit 7.x on PHP 7.1

swissspidy committed:
Fix string interpolation in `CachedRemoteGetRequestTest`

swissspidy committed:
Avoid passing `null` to `parse_str`

swissspidy committed:
Add `ReturnTypeWillChange` attribute in `AMP_Validation_Callback_Wrapper`

swissspidy committed:
Do not install `pcov/clobber` for PHPUnit 8

swissspidy committed:
Avoid passing `null` to `parse_str`, take 2

swissspidy committed:
Add missing attribute

swissspidy committed:
Fix another string interpolation issue

westonruter committed:
Update amp-toolbox-php to 0.11.3-alpha

westonruter committed:
Merge remote-tracking branch 'origin/develop' into try-php81-php82

* origin/develop: (32 commits)
  Bump classnames from 2.3.1 to 2.3.2
  Bump webpack-remove-empty-scripts from 1.0.0 to 1.0.1
  Bump wp-cli/extension-command from 2.1.6 to 2.1.7
  Bump postcss-preset-env from 7.8.1 to 7.8.2
  Update jest snapshots after @wordpress/components package updates
  Update Gutenberg package dependencies
  Bump sirbrillig/phpcs-variable-analysis from 2.11.7 to 2.11.8
  Fix failing tests on trunk due to webp support
  Remove avoidable isset() check
  Bump @babel/core from 7.18.13 to 7.19.0
  Bump eslint from 8.23.0 to 8.23.1
  Bump eslint-plugin-jest from 27.0.1 to 27.0.4
  Bump eslint-plugin-react from 7.31.1 to 7.31.8
  Bump webpack-remove-empty-scripts from 0.8.1 to 1.0.0
  Bump css-minimizer-webpack-plugin from 4.0.0 to 4.1.0
  Update Gutenberg package dependencies
  Bump postcss-preset-env from 7.8.0 to 7.8.1
  Update test snapshots
  Update mask-type attribute with mask-type css prop
  Bump uuid from 8.3.2 to 9.0.0
  ...

westonruter commented on amp.php:
```diff
@@ -257,3 +257,5 @@ function _amp_incorrect_plugin_slug_admin_notice() {
 register_deactivation_hook( __FILE__, 'amp_deactivate' );
 
 add_action( 'plugins_loaded', 'amp_bootstrap_plugin', defined( 'PHP_INT_MIN' ) ? PHP_INT_MIN : ~PHP_INT_MAX ); // phpcs:ignore PHPCompatibility.Constants.NewConstants.php_int_minFound
+
+// TEMP TO TRIGGER CI. TODO: REMOVE.
```
Can be removed now that PHP files have changed.

```suggestion
```

swissspidy committed:
Remove temp change

Co-authored-by: Weston Ruter <westonruter@google.com>

swissspidy commented:
Something to fix in PHP-CSS-Parser:

```
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 129
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 129
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 130
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 131
```

thelovekesh commented on .github/workflows/build-test-measure.yml:
```diff
@@ -491,12 +534,6 @@ jobs:
         if: needs.pre-run.outputs.changed-php-count > 0
         run: bash bin/ci/after-wp-install.sh ${{ matrix.wp }} ${{ matrix.install-pwa-plugin == true }}
 
-      - name: Setup PCOV
-        if: ${{ matrix.coverage == true && needs.pre-run.outputs.changed-php-count > 0 }}
-        run: |
-          composer require --dev --ignore-platform-reqs pcov/clobber
-          vendor/bin/pcov clobber
-
```
I think pcov is disabled temporarily, right?

swissspidy replied:
PCOV support was merged into PHPUnit 8. `pcov/clobber` is no longer needed.

westonruter commented on tests/php/src/Instrumentation/EventTest.php:
```diff
@@ -84,7 +84,6 @@ public function data_add_properties_typing() {
 			'bool key'                   => [ [ true => 'value' ], InvalidEventProperties::class, 'but found an element key of type integer' ],
 			'string key'                 => [ [ 'key' => 'value' ], false ],
 			'int key'                    => [ [ 42 => 'value' ], InvalidEventProperties::class, 'but found an element key of type int' ],
-			'float key'                  => [ [ 3.14 => 'value' ], InvalidEventProperties::class, 'but found an element key of type integer' ],
```
See https://github.com/ampproject/amp-wp/commit/924be547674068a54c62fb87fcc207f03f6af27f

> Using floats as array keys is deprecated in PHP 8.1

westonruter commented on tests/php/bootstrap.php:
```diff
@@ -59,10 +59,13 @@ function amp_unit_test_load_plugin_file() {
 
-$logger = new WP_CLI\Loggers\Regular( true );
-WP_CLI::set_logger( $logger );
+if ( file_exists( WP_CLI_ROOT . '/php/utils.php' ) ) {
```
Why was the `file_exists()` check needed? Is it due to WP-CLI not being able to be installed on PHP 8.2?

swissspidy replied:
Not 100% sure right now (not at computer) but yeah I think it was because of issues installing it. Maybe because of Requests? Will have to check it out again.

swissspidy replied:
Here's the run before that change: https://github.com/ampproject/amp-wp/actions/runs/2925960305/jobs/4667211482

```
PHP Warning:  require_once(/tmp/wordpress/src/wp-content/plugins/amp/vendor/wp-cli/wp-cli/php/utils.php): Failed to open stream: No such file or directory in /tmp/wordpress/src/wp-content/plugins/amp/tests/php/bootstrap.php on line 62
```

westonruter commented on includes/validation/class-amp-validation-callback-wrapper.php:
```diff
@@ -393,6 +393,7 @@ protected function finalize_scripts( WP_Scripts $wp_scripts, array $before_regis
 	 * @param mixed $offset Offset.
 	 * @param mixed $value  Value.
 	 */
+	#[\ReturnTypeWillChange]
```
I suppose this is needed because we can't specify the matching return type since return types aren't available in older versions of PHP. 

Oh, but actually we now require PHP 7 in the AMP plugin. So instead of this attribute could we instead just add the return types?

swissspidy replied:
If we‘re not worried about someone extending the class and breaking their implementation because of the return types, sure.

swissspidy replied:
See https://php.watch/versions/8.1/ReturnTypeWillChange

westonruter replied:
No, I'm not worried about someone extending the class. It should have been marked `final` to begin with.

swissspidy replied:
Since we can't use `void` return types yet (they're a PHP 7.1 feature), can't really change this.

westonruter commented:
There's currently an error for PHP 7.4 with coverage.

swissspidy commented on phpunit.xml.dist:
```diff
@@ -2,10 +2,10 @@
 	bootstrap="tests/php/bootstrap.php"
 	backupGlobals="false"
 	colors="true"
-	convertDeprecationsToExceptions="true"
-	convertErrorsToExceptions="true"
-	convertNoticesToExceptions="true"
-	convertWarningsToExceptions="true"
```
This can be reverted again once there are no warnings from WP core anymore (looking at you, Requests library)

swissspidy commented on src/Dom/ElementList.php:
```diff
@@ -35,6 +35,7 @@ final class ElementList implements IteratorAggregate, Countable {
 	 * @param DOMElement|null $caption The caption for the element.
 	 * @return ElementList A clone of this list, with the new element added.
 	 */
+	#[\ReturnTypeWillChange]
```
I think here we can safely add the return types because it's a final class

swissspidy committed:
Add return type instead of attribute

swissspidy committed:
Use xdebug instead of pcov

swissspidy committed:
Update `testVersion` in PHPCS config since the plugin requires 7.0+

swissspidy committed:
Address unrelated PHPCS warnings

```
Redeclaration of variable $cache_hit_callback as static variable. (VariableAnalysis.CodeAnalysis.VariableAnalysis.VariableRedeclaration)
```

swissspidy commented:
> There's currently an error for PHP 7.4 with coverage.

Yeah I saw it too and it's pretty confusing.

Somehow the test never finishes, so there's no coverage file being created:

https://github.com/ampproject/amp-wp/actions/runs/3087716934/jobs/5021640946#step:21:55

A couple of PHP warnings:

```
PHP Warning:  Cannot modify header information - headers already sent by (output started at /tmp/wordpress-tests-lib/includes/bootstrap.php:261) in /tmp/wordpress/src/wp-includes/pluggable.php on line 1421

Warning: Cannot modify header information - headers already sent by (output started at /tmp/wordpress-tests-lib/includes/bootstrap.php:261) in /tmp/wordpress/src/wp-includes/pluggable.php on line 1424
```

Then I noticed that while CI is supposed to use PCOV for coverage ever since https://github.com/ampproject/amp-wp/commit/536dbe176596c375becaeadd1151833d9c7353e9, it has always been using XDebug.

See https://github.com/ampproject/amp-wp/actions/runs/2749592396/jobs/4314380797#step:20:17 from 2 months ago.

Note how it says "Runtime:       PHP 7.4.30 with Xdebug"

So it was never using PCOV at all.

Now I've tried switching to XDebug again but it doesn't work either. Same error though 😞 

No idea why. `wp-includes/pluggable.php on line 1424` is the `wp_redirect()` function. So something is doing redirects when trying to install WordPress...

swissspidy commented on .phpcs.xml.dist:
```diff
@@ -112,7 +112,7 @@
 
 	<!-- Include sniffs for PHP cross-version compatibility. -->
-	<config name="testVersion" value="5.6-"/>
+	<config name="testVersion" value="7.0-"/>
```
The plugin requires PHP 7+

Without this we're getting yelled at by PHPCS for using return type declarations.

westonruter replied:
Oh, good catch. This should have been updated as part of #7088.

swissspidy committed:
Make actual use of `$_tests_dir` in test bootstrap

swissspidy committed:
Merge branch 'develop' into try-php81-php82

swissspidy committed:
Merge branch 'develop' into try-php81-php82

# Conflicts:
#	composer.lock

swissspidy committed:
Try running coverage on PHP 8.0

swissspidy committed:
Fix `test_handle_single_url_page_bulk_and_inline_actions` test

swissspidy committed:
Fix `test_prepare_response_throwing_error` on trunk

swissspidy commented:
@westonruter I found the issue 🥳 🎉 

The culprit was the `test_handle_single_url_page_bulk_and_inline_actions` test covering `\AMP_Validation_Error_Taxonomy::handle_single_url_page_bulk_and_inline_actions()`. That method calls `wp_safe_redirect` and `exit()`, causing the tests to silently halt.

I update the test to short-circuit the redirect and wrapping the `exit` in an `if` so that it's never called during tests.

While at it, I noticed that `test_prepare_response_throwing_error` failed on trunk because on trunk there are currently tons of other PHP deprecation warnings, making the regex not match correctly. Making the regex less strict resolves the issue.

All tests should be passing now.

swissspidy committed:
Fix automatic conversion of false to array

swissspidy committed:
Fix `test_prepare_response_throwing_exception`

swissspidy commented:
Slight correction, there are two more tests failing on trunk. They are unrelated to the changes here though.

```
1) AMP_Style_Sanitizer_Test::test_prioritized_stylesheets with data set "admin_bar_included" (Closure Object (...), Closure Object (...))
Failed asserting that '' contains "admin-bar".

Error: /tmp/wordpress/src/wp-content/plugins/amp/tests/php/test-amp-style-sanitizer.php:3494
/tmp/wordpress/src/wp-content/plugins/amp/tests/php/test-amp-style-sanitizer.php:3575

2) Test_AMP_Theme_Support::test_prepare_response_standard
'<link rel="preload" as="script" href="https://cdn.ampproject.org/v0/amp-dynamic-css-classes-0.1.js">' is not after '<link rel="dns-prefetch" href="//cdn.ampproject.org">'
Failed asserting that 364 is greater than 61757.

/tmp/wordpress/src/wp-content/plugins/amp/tests/php/test-class-amp-theme-support.php:1767

```

Both fail because the HTML output is polluted with deprecation warnings like this from core:

```
Deprecated: strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated in /tmp/wordpress/wp-content/plugins/gutenberg/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php on line 1113
```

This bug was just fixed in Gutenberg yesterday: https://github.com/WordPress/gutenberg/pull/44829
So this will get resorted once the next GB version is released.

swissspidy commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -265,10 +265,14 @@ public static function init() {
+	 * @return array|false Filtered options.
 	 */
 	public static function filter_options_when_force_standard_mode_request( $options ) {
+		if ( ! $options ) {
+			$options = [];
+		}
+
```
Prevents deprecation warnings for automatic conversion of `false` to `array`

thelovekesh commented on .github/workflows/build-test-measure.yml:
```diff
@@ -425,8 +434,7 @@ jobs:
           extensions: curl, date, dom, gd, iconv, json, libxml, mysql, spl
-          coverage: ${{ matrix.coverage && 'pcov' || 'none' }}
-          ini-values: pcov.directory=.
+          coverage: ${{ matrix.coverage && 'xdebug' || 'none' }}
```
Should we continue using `pcov` for coverage in the CI? I've looked at the problem that it uses xdebug in any case, but I can debug it in another PR. Thoughts?

swissspidy replied:
I'll defer to Weston, but if it works and is faster, sure.

But since this never really worked in the first place, definitely warrants its own issue/PR.

thelovekesh replied:
PCOV is faster but lacks path coverage. A difference in reporting can be seen [here](https://github.com/krakjoe/pcov#differences-in-reporting).

westonruter replied:
PCOV was introduced in https://github.com/ampproject/amp-wp/pull/5201

I just pinged you both the thread where this was discussed.

thelovekesh commented on composer.json:
```diff
@@ -14,7 +14,7 @@
     "ext-libxml": "*",
     "ext-spl": "*",
-    "ampproject/amp-toolbox": "0.11.2",
+    "ampproject/amp-toolbox": "dev-main",
```
@westonruter Is it intended to bump it to `dev-main`?

westonruter replied:
Yes, in order to update php-toolbox-php to the version that adds PHP 8.2 support: https://github.com/ampproject/amp-toolbox-php/pull/533

thelovekesh commented on phpunit.xml.dist:
```diff
@@ -2,10 +2,10 @@
-	convertErrorsToExceptions="true"
-	convertNoticesToExceptions="true"
-	convertWarningsToExceptions="true"
+	convertDeprecationsToExceptions="false"
+	convertErrorsToExceptions="false"
+	convertNoticesToExceptions="false"
+	convertWarningsToExceptions="false"
```
It can be reverted now?

swissspidy replied:
Not yet. See https://github.com/ampproject/amp-wp/pull/7225#discussion_r977290405

thelovekesh reviewed (approved):
Aside from the tests that failed and will be fixed following the new GB release, the rest LGTM 👍🏼

swissspidy committed:
Merge branch 'develop' into try-php81-php82

swissspidy commented:
> This bug was just fixed in Gutenberg yesterday: [WordPress/gutenberg#44829](https://github.com/WordPress/gutenberg/pull/44829)
> So this will get resorted once the next GB version is released.

FYI, while GB 14.4 was released _today_, that bugfix won't be included until version 14.5 ([due by November 07, 2022](https://github.com/WordPress/gutenberg/milestone/190))

swissspidy commented:
> Unit test (with coverage): PHP 7.4, WP latest

This required status check needs to be removed/changed, since now the coverage is done with PHP 8.0 instead.

westonruter commented:
> > Unit test (with coverage): PHP 7.4, WP latest
> 
> This required status check needs to be removed/changed, since now the coverage is done with PHP 8.0 instead.

Removed (to be changed after merge)

westonruter commented on includes/validation/class-amp-validation-manager.php:
```diff
@@ -265,10 +265,14 @@ public static function init() {
-	 * @param array $options Options.
-	 * @return array Filtered options.
+	 * @param array|false $options Options.
+	 * @return array|false Filtered options.
```
Isn't the return value now always made an array?

```suggestion
	 * @return array Filtered options.
```

westonruter commented on includes/validation/class-amp-validation-error-taxonomy.php:
```diff
@@ -2885,13 +2885,14 @@ public static function handle_single_url_page_bulk_and_inline_actions( $post_id
-		);
-		exit();
+		) ) {
+			exit();
```
```suggestion
			exit(); // @codeCoverageIgnore
```

westonruter commented on src/Admin/SiteHealth.php:
```diff
@@ -534,6 +534,16 @@ public function get_page_cache_detail( $use_previous_result = false ) {
+	 * @param string $header_value Header value.
+	 * @return bool Whether header contains the value "hit"
+	 */
+	private static function cache_hit_callback( string $header_value ): bool {
```
This was introduced in bd2ee42bf5c035ad1d3868d6f40400d6af13d99d apparently to deal with a PHPCS bug, and not due to any fault of the code?

swissspidy replied:
Correct

westonruter commented on tests/php/bootstrap.php:
```diff
@@ -7,27 +7,14 @@
 
 define( 'TESTS_PLUGIN_DIR', dirname( dirname( __DIR__ ) ) );
 
-// Detect where to load the WordPress tests environment from.
-if ( false !== getenv( 'WP_TESTS_DIR' ) ) {
-	$_test_root = getenv( 'WP_TESTS_DIR' );
-} elseif ( false !== getenv( 'WP_DEVELOP_DIR' ) ) {
-	$_test_root = getenv( 'WP_DEVELOP_DIR' ) . '/tests/phpunit';
-} elseif ( file_exists( '/tmp/wordpress-tests/includes/bootstrap.php' ) ) {
-	$_test_root = '/tmp/wordpress-tests';
-} elseif ( file_exists( '/var/www/wordpress-develop/tests/phpunit' ) ) {
-	$_test_root = '/var/www/wordpress-develop/tests/phpunit';
-} else {
-	$_test_root = dirname( dirname( dirname( dirname( TESTS_PLUGIN_DIR ) ) ) ) . '/tests/phpunit';
-}
-
 // When run in wp-env context, set the test config file path.
 if ( ! defined( 'WP_TESTS_CONFIG_FILE_PATH' ) && false !== getenv( 'WP_PHPUNIT__TESTS_CONFIG' ) ) {
 	define( 'WP_TESTS_CONFIG_FILE_PATH', getenv( 'WP_PHPUNIT__TESTS_CONFIG' ) );
 }
 
 require_once TESTS_PLUGIN_DIR . '/vendor/yoast/wp-test-utils/src/WPIntegration/bootstrap-functions.php';
 $_tests_dir = Yoast\WPTestUtils\WPIntegration\get_path_to_wp_test_dir();
-require_once $_test_root . '/includes/functions.php';
```
Nice catch

swissspidy committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter reviewed (approved):
Thanks for doing this! 

Shall we wait to merge until Gutenberg 14.5 is released?

swissspidy commented:
I don‘t see an issue with merging this already, but up to you

westonruter commented:
The issue is the failing tests will make it hard to identify problems with other PRs being opened in the next ~2 weeks.

swissspidy commented:
You can also just temporarily skip these two tests

westonruter committed:
Merge branch 'develop' into try-php81-php82

westonruter commented:
Gutenberg updated in https://github.com/ampproject/amp-wp/pull/7331. PHPUnit tests now all pass. 🎉

pavanpatil1 commented:
Tested it with PHP 8.1 and 8.2, Plugin functionality seems working fine with both the PHP versions. However, there are some deprecation warnings are visible in both versions. 

**PHP8.1**

```php
PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in C:\Users\win 10\Local Sites\php8amp\app\public\wp-content\plugins\amp\includes\class-amp-http.php on line 95
```


**PHP8.2**

```php
PHP Deprecated:  Return type of AMP_Validation_Callback_Wrapper::offsetUnset($offset) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/htdocs/current/wp-content/plugins/amp/includes/validation/class-amp-validation-callback-wrapper.php on line 425
PHP Deprecated:  Creation of dynamic property AmpProject\Optimizer\Configuration\MinifyHtmlConfiguration::$minifyAmpScript is deprecated in /var/www/htdocs/current/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Optimizer/Configuration/BaseTransformerConfiguration.php on line 40
PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in /var/www/htdocs/current/wp-content/plugins/amp/includes/class-amp-http.php on line 92
```

westonruter commented:
In regards to:

```
PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in C:\Users\win 10\Local Sites\php8amp\app\public\wp-content\plugins\amp\includes\class-amp-http.php on line 95
```

This is the code in question:

https://github.com/ampproject/amp-wp/blob/68d9bb3f977967fa05fb09f5f52da561dc85bd72/includes/class-amp-http.php#L78-L98

It looks like all that is required is:

```diff
- 				'status_code' => null,
+ 				'status_code' => 0,
```

westonruter commented:
In regards to:

```
PHP Deprecated:  Return type of AMP_Validation_Callback_Wrapper::offsetUnset($offset) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/htdocs/current/wp-content/plugins/amp/includes/validation/class-amp-validation-callback-wrapper.php on line 425
```

I'm confused since it _does_ have `#[\ReturnTypeWillChange]`:

https://github.com/ampproject/amp-wp/blob/68d9bb3f977967fa05fb09f5f52da561dc85bd72/includes/validation/class-amp-validation-callback-wrapper.php#L422-L433

westonruter commented:
And finally, in regards to:

```
PHP Deprecated:  Creation of dynamic property AmpProject\Optimizer\Configuration\MinifyHtmlConfiguration::$minifyAmpScript is deprecated in /var/www/htdocs/current/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Optimizer/Configuration/BaseTransformerConfiguration.php on line 40
```

Looks like this is something that was missed in https://github.com/ampproject/amp-toolbox-php/pull/533. 

See example at https://github.com/ampproject/amp-toolbox-php/pull/533/files#diff-12262a92eb1759cfc297901a94d3e5ab9edc9862154fb003e21f7b80719a7fc2

westonruter commented:
Actually, it seems like this was tested without the latest `amp-toolbox-php`. Line 40 in `BaseTransformerConfiguration.php` is the [following in v0.11.2](https://github.com/ampproject/amp-toolbox-php/blob/8fa158845c980437831397100ca7133eee76c61d/src/Optimizer/Configuration/BaseTransformerConfiguration.php#L40):

```php
$this->$key = $this->validate($key, $value);
```

Whereas in `main` it is the [following](https://github.com/ampproject/amp-toolbox-php/blob/bd739ddcc839a70634bfefe012cbcd2268b51f7e/src/Optimizer/Configuration/BaseTransformerConfiguration.php#L40):

```php
$this->allowedKeys = $this->getAllowedKeys();
```

I just tried downloading the most recent development build and I see that the former is present as expected.

@pavanpatil1 Could you please re-test, making sure you have the latest [development build](https://github.com/ampproject/amp-wp/wiki/Development-Builds) installed (as for all current QA)?

thelovekesh commented:
I checked with PHP 8.2 but was unable to get any deprecation warnings other than the
```bash
[06-Feb-2023 19:38:52 UTC] PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in /app/public/content/plugins/amp/includes/class-amp-http.php on line 92
```

pavanpatil1 commented:
QA passed ✅ 

Tested it with the latest [development build](https://storage.googleapis.com/ampwp_github_artifacts/refs/heads/develop/dev/amp.zip). Functionality was working as expected before now the deprecation warnings are not visible now with both v8.1 and 8.2. solved in [this](https://github.com/ampproject/amp-wp/pull/7452) PR.
